### ISSUE TITLE
asgd in mpi-sgd 

### ISSUE THREAD

**Issue by slowbull** [raingo/caffe-mpi#1]
Your sgd-mpi implementation is a littlle confusing.  It seems like that each worker uses parameters updated by itself last time, instead of pulling current parameters in the server. 
 I think it should import more bias in the computation, because its time delay is larger.

**Comment by raingo** [comment]
Each worker will send the diff to root node, and wait for update data.

Please refer to the https://github.com/raingo/caffe-mpi/blob/master/sgd-mpi.cpp#L57-L65

**Comment by fastalgo** [comment]
**I compiled the code, and got errors like the following.**

In file included from sgd.cpp:4:0:
evaluator.hpp: In function ‘void evaluate(boost::shared_ptr<caffe::Net<float> >, int)’:
evaluator.hpp:89:46: error: no matching function for call to ‘caffe::Net<float>::ForwardBackward(std::vector<caffe::Blob<float>*, std::allocator<caffe::Blob<float>*> >&)’
             net -> ForwardBackward(bottom_vec);
                                              ^
evaluator.hpp:89:46: note: candidate is:
In file included from /opt/intel/intelcaffe/include/caffe/caffe.hpp:49:0,
                 from common.hpp:12,
                 from sgd.cpp:2:
/opt/intel/intelcaffe/include/caffe/net.hpp:124:9: note: Dtype caffe::Net<Dtype>::ForwardBackward() [with Dtype = float]
   Dtype ForwardBackward() {
         ^
/opt/intel/intelcaffe/include/caffe/net.hpp:124:9: note:   candidate expects 0 arguments, 1 provided
sgd.cpp: In function ‘int main(int, char**)’:
sgd.cpp:33:51: error: no matching function for call to ‘caffe::Net<float>::ForwardBackward(std::vector<caffe::Blob<float>*, std::allocator<caffe::Blob<float>*> >&)’
     Dtype loss = net -> ForwardBackward(bottom_vec);
                                                   ^
sgd.cpp:33:51: note: candidate is:
In file included from /opt/intel/intelcaffe/include/caffe/caffe.hpp:49:0,
                 from common.hpp:12,
                 from sgd.cpp:2:
/opt/intel/intelcaffe/include/caffe/net.hpp:124:9: note: Dtype caffe::Net<Dtype>::ForwardBackward() [with Dtype = float]
   Dtype ForwardBackward() {
         ^
/opt/intel/intelcaffe/include/caffe/net.hpp:124:9: note:   candidate expects 0 arguments, 1 provided
sgd.cpp:40:55: error: no matching function for call to ‘caffe::Net<float>::ForwardBackward(std::vector<caffe::Blob<float>*, std::allocator<caffe::Blob<float>*> >&)’
         Dtype loss = net -> ForwardBackward(bottom_vec);
                                                       ^
sgd.cpp:40:55: note: candidate is:
In file included from /opt/intel/intelcaffe/include/caffe/caffe.hpp:49:0,
                 from common.hpp:12,
                 from sgd.cpp:2:
/opt/intel/intelcaffe/include/caffe/net.hpp:124:9: note: Dtype caffe::Net<Dtype>::ForwardBackward() [with Dtype = float]
   Dtype ForwardBackward() {
         ^
/opt/intel/intelcaffe/include/caffe/net.hpp:124:9: note:   candidate expects 0 arguments, 1 provided
make: *** [.sgd.o] Error 1
